### ISSUE TITLE
Why is the detection results of val.py better than that of detect.py.   Using yolov5 for defect detection, val.py can detect, but detect.py can't.

### ISSUE THREAD

**Issue by BossCrab-jyj** [ultralytics/yolov5#6745]
Why is the detection results of val.py better than that of detect.py.   Using yolov5 for defect detection, val.py can detect, but detect.py can't.   

I found that during training, the input size of the model is [416,416]. When using val.py for evaluation, the input size of the model is [352,448]. This should be the size after transformation, but when I use detect.py for detection, the input model size is [352,416], and the result is much worse than that of val.py. Then after I change the model input size of  [352,448] manually, the result will be better than [352,416], but the result is still not as good as val.py, which makes me very confused.

_Originally posted by @yingjie-jiang in https://github.com/ultralytics/yolov5/issues/4720#issuecomment-1048542361_

**Comment by supriamir** [comment]
Hi @glenn-jocher 
is it mean that the result between val.py and inference will never be same?
I have tried to use the inference same parameter setting with val.py, the result of precision and recall are different.

Here is my setting
model = torch.hub.load('.', 'custom', pathmodel, source='local')
model.conf = 0.25 
iou_thres = 0.5

Thank you

**Comment by glenn-jocher** [comment]
Hi @supriamir,

The results between `val.py` and `detect.py` can vary due to several factors. One important thing to note is that `val.py` is designed for evaluation during training, while `detect.py` is meant for inference on new images.

During training, `val.py` uses the validation dataset and performs a thorough evaluation of the model's performance. It applies transformations such as scaling and cropping to the input images for better generalization. On the other hand, `detect.py` focuses on real-time inference on new images, where the input size may not undergo the same transformations as during training.

In your case, the differences in input size ([352,448] for `val.py` vs [352,416] for `detect.py`) and the manual adjustment you made could affect the detection results. Other factors, such as the configuration of the model's `conf` (confidence threshold) and `iou_thres` (IoU threshold), can also influence the output.

It's worth mentioning that the precision and recall metrics themselves may differ slightly between `val.py` and `detect.py` due to the nature of these evaluation scenarios. It's recommended to evaluate the model's performance using a consistent approach, such as utilizing `val.py`, which provides a comprehensive assessment during training.

If you have further questions or need assistance with specific parameters or usage, feel free to ask. The YOLO community and the Ultralytics team are always here to help.

Best, Glenn

**Comment by supriamir** [comment]
@glenn-jocher thank you for answer.

1. Yolov5 doesn't provide the metrics based on the scales, so I used the detect.py for calculate the precision and recall based on the scales (small, medium, large). 

2. For the yolov8, is there a way to print the results using pandas? In yolov5, we can use  model(img_path).pandas().xyxy[0]

**Comment by glenn-jocher** [comment]
Hi @supriamir,

You're welcome! 

1. YOLOv5 doesn't provide metrics specifically based on the scales (small, medium, large), so using `detect.py` to calculate precision and recall for each scale can be a reasonable approach. Keep in mind that these metrics may not directly align with the default evaluation metrics used in `val.py`.

2. YOLOv8 doesn't have a built-in method to directly print results using pandas like `model(img_path).pandas().xyxy[0]` in YOLOv5. However, you can still extract the detection results in YOLOv8 and convert them to a pandas DataFrame if desired. You can use the `torchvision` post-processing functions provided by YOLOv8 to retrieve the detections, and then convert them to a pandas DataFrame using pandas' `DataFrame` constructor.

Please let me know if you have any further questions or need more assistance.

Regards,
Glenn

**Comment by supriamir** [comment]
@glenn-jocher Thank you.

**Comment by glenn-jocher** [comment]
Hi @supriamir,

You're welcome! If you have any more questions or need further assistance, feel free to ask. We're here to help!

Thank you,
Glenn

**Comment by supriamir** [comment]
Hi @glenn-jocher 

I want to use the custom model and run the inference on my local machine. Is yolov8 going to provide the function like yolov5 does model= torch.hub.load('.', 'custom', pathmodel, source='local')

**Comment by glenn-jocher** [comment]
Hi @supriamir,

Yes, in YOLOv8, you can also load a custom model for inference on your local machine. However, the specific syntax and methods may differ from YOLOv5.

For YOLOv8, you'll need to follow the documentation or code examples provided by the YOLOv8 repository to load and run your custom model. Make sure to check the repository's README or other relevant documentation for instructions on how to load and use custom models.

If you have any further questions or need assistance with the YOLOv8 implementation, please feel free to ask.

Best, Glenn

**Comment by supriamir** [comment]
Hi @glenn-jocher 

I have a couple of questions related to YOLOv5:
1. What method YOLOv5 use to calculate the Average Precision (AP)? 
2. I got low mAP than precision and recall. is it normal? Example : Precision = 0.6755 , Recall = 0.7710 and mAP = 0.6684

Looking forward to your response.

Thank you.

**Comment by glenn-jocher** [comment]
Hi @supriamir,

Regarding your questions on YOLOv5:

1. YOLOv5 calculates the Average Precision (AP) using the standard method used in object detection. It computes the precision and recall at different confidence thresholds, and then calculates the area under the precision-recall curve to determine the AP. This metric provides an overall assessment of the model's detection performance.

2. It's not uncommon to have a slightly lower mAP than the precision and recall values. This can happen due to variations in the detection accuracy across different object categories or difficulty levels. The mAP is an average value that considers the performance across all object categories, which may bring down the overall score if there are specific categories with lower accuracy.

If you have any more questions or need further clarification, please don't hesitate to ask. We're here to help.

Glenn

**Comment by supriamir** [comment]
@glenn-jocher thank you for your response.

1. Do you mean calculates the area under the precision-recall curve with 11-interpolated precision?

**Comment by supriamir** [comment]
Hi @glenn-jocher 

I have a couple of questions related to CIOU, DIoU:

1. I try to use DIoU instead of CIoU. I have changed in the file utils/metrics.py (DIOU=true). Should I also need to change the loss.py because the default setting is CIoU=True ?
                    iou = bbox_iou(pbox, tbox[i], CIoU=True).squeeze()  # iou(prediction, target) -->here?

2. I calculate the AP using the inference not from val.py so should I also use the formula of DIOU to calcuate the IOU?

change the default formula ` iou = (inter_area) / (pred_area + gt_area - inter_area)`  to 

`iou = inter_area / (pred_area + gt_area - inter_area)
                center_distance = ((results.loc[i, 'xmin'] + results.loc[i, 'xmax']) / 2 - ground_truth[1])**2 + ((results.loc[i, 'ymin'] + results.loc[i, 'ymax']) / 2 - ground_truth[2])**2
                diagonal_distance = (gt_xmax - gt_xmin)**2 + (gt_ymax - gt_ymin)**2
                diou = iou - center_distance / diagonal_distance`
 

Looking forward to your response.

Thank you.

**Comment by glenn-jocher** [comment]
Hi @supriamir,

Regarding your questions on CIOU and DIOU:

1. If you have changed the CIoU setting in `utils/metrics.py` to DIOU=True, then you should also update the corresponding calculation in `loss.py`. Ensure that the calculation for `iou` in `loss.py` is consistent with the setting you have chosen. Changing `iou = bbox_iou(pbox, tbox[i], CIoU=True).squeeze()` to `iou = bbox_iou(pbox, tbox[i], CIoU=False).squeeze()` will ensure the DIOU calculation is used in the loss function.

2. If you're calculating the Average Precision (AP) using inference instead of `val.py`, it would be necessary to use the DIOU formula to calculate the IOU. The formula you provided for DIOU calculation is correct. By incorporating the center distance and diagonal distance, you can calculate the DIOU as `diou = iou - center_distance / diagonal_distance`. Please ensure that you update the calculation of IOU in your evaluation code accordingly.

If you have any further questions or need additional clarification, feel free to ask. We're here to support you.

Glenn

**Comment by supriamir** [comment]
Hi @glenn-jocher 

Thank you your answer. I've been working on calculating metrics such as Precision, Recall, and Average Precision (AP) using the formulations of CIOU, DIoU, and GIOU as implemented in the code from the metrics.py file. Here's the specific code snippet I'm referring to:

                    eps = 1e-7
                    # Intersection area
                    inter = (b1_x2.minimum(b2_x2) - b1_x1.maximum(b2_x1)).clamp(0) * \
                            (b1_y2.minimum(b2_y2) - b1_y1.maximum(b2_y1)).clamp(0)

                    # Union Area
                    union = w1 * h1 + w2 * h2 - inter + eps

                    # IoU
                    iou = inter / union

                    if metric_type == 'CIoU' or metric_type == 'DIoU' or metric_type == 'GIoU':
                        cw = b1_x2.maximum(b2_x2) - b1_x1.minimum(b2_x1)  # convex (smallest enclosing box) width
                        ch = b1_y2.maximum(b2_y2) - b1_y1.minimum(b2_y1)  # convex height
                        c2 = cw ** 2 + ch ** 2 + eps  # convex diagonal squared
                        rho2 = ((b2_x1 + b2_x2 - b1_x1 - b1_x2) ** 2 + (b2_y1 + b2_y2 - b1_y1 - b1_y2) ** 2) / 4  # center dist ** 2
                        if metric_type == 'CIoU':
                            v = (4 / math.pi ** 2) * (torch.atan(w2 / h2) - torch.atan(w1 / h1)).pow(2)
                            alpha = v / (v - iou + (1 + eps))
                            iou_metric = iou - (rho2 / c2 + v * alpha)  # CIoU
                        elif metric_type == 'DIoU':
                            iou_metric = iou - rho2 / c2  # DIoU
                        elif metric_type == 'GIoU':
                            c_area = cw * ch + eps  # convex area
                            iou_metric = iou - (c_area - union) / c_area  # GIoU
                    else:
                        iou_metric = iou


I'm currently seeking clarification on whether this code accurately calculates the desired metrics. If you have any insights, guidance, or assistance to offer regarding the accuracy of this code, I would greatly appreciate your input.

**Comment by glenn-jocher** [comment]
Hi @supriamir,

Yes, the code snippet you provided from the `metrics.py` file correctly implements the CIOU, DIoU, and GIOU metrics for calculating metrics such as Precision, Recall, and Average Precision (AP) in YOLOv5. The formulas for these metrics, incorporating the intersection area, union area, and necessary adjustments for the different metric types, are accurately implemented in the code snippet you shared.

If you have any further questions or need additional assistance, feel free to ask. We're here to help!

Glenn

**Comment by supriamir** [comment]
Hi @glenn-jocher 

Can we run yolov3 in Ultralytics/YOLOv8?



    # Load a model
    model = YOLO("ultralytics/cfg/models/v3/yolov3-spp.yaml")  # build a new model from scratch
    # Use the model
    model.train(data="uav.yaml", epochs=300,imgsz=640, batch=8, patience=0, workers=1)  # train the model
    metrics = model.val()

**Comment by glenn-jocher** [comment]
Hi @supriamir,

In Ultralytics YOLOv8, it's not possible to directly run YOLOv3 models using the provided syntax. YOLOv8 is specifically designed to work with YOLOv5 models and may not be compatible with YOLOv3 models out of the box.

If you want to use YOLOv3 models, you may need to use the original YOLOv3 implementation or a framework that supports YOLOv3 models directly.

If you have any further questions or need assistance with YOLOv3 or any other related queries, please feel free to ask. We're here to support you.

Glenn
